File: 30-substring_with_concatenation_of_all_words.js
var findSubstring = function(s, words) {
    // Edge case: If no words are provided, return empty array
    if (words.length === 0 || s.length === 0) return [];

    const wordLength = words[0].length;
    const wordCount = words.length;
    const substringLength = wordLength * wordCount;
    const wordMap = new Map();

    // Populate the map with the frequency of each word in words
    for (let word of words) {
        wordMap.set(word, (wordMap.get(word) || 0) + 1);
    }

    const result = [];

    // Loop through each possible starting position of the substring
    for (let i = 0; i < wordLength; i++) {
        let left = i;
        let right = i;
        let currentMap = new Map();
        let count = 0;

        // Sliding window approach to check substrings
        while (right + wordLength <= s.length) {
            // Get the word at the current right index
            const word = s.slice(right, right + wordLength);
            right += wordLength; // Move the right pointer forward

            // If the word is part of the words list
            if (wordMap.has(word)) {
                // Update the currentMap and count for valid words
                currentMap.set(word, (currentMap.get(word) || 0) + 1);
                count++;

                // If we have more than required count of this word, move the left pointer
                while (currentMap.get(word) > wordMap.get(word)) {
                    const leftWord = s.slice(left, left + wordLength);
                    currentMap.set(leftWord, currentMap.get(leftWord) - 1);
                    left += wordLength;
                    count--;
                }

                // If we have seen exactly wordCount valid words, add the index to result
                if (count === wordCount) {
                    result.push(left);
                    const leftWord = s.slice(left, left + wordLength);
                    currentMap.set(leftWord, currentMap.get(leftWord) - 1);
                    left += wordLength;
                    count--;
                }
            } else {
                // If the word is not in the list, reset the sliding window
                currentMap.clear();
                count = 0;
                left = right;
            }
        }
    }

    return result;
};

File: 30-substring_with_concatenation_of_all_words.py
from collections import Counter

def findSubstring(s: str, words: list) -> list:
    if not s or not words or not words[0]:  # Edge case check
        return []
    
    word_length = len(words[0])  # Length of each word (all are the same length)
    word_count = len(words)  # Number of words
    substring_length = word_length * word_count  # Total length of the concatenated substring
    word_freq = Counter(words)  # Count occurrences of each word
    result = []  # List to store valid starting indices
    
    # Iterate over possible starting points within a word-length window
    for i in range(word_length):
        left = i  # Start of the current window
        right = i  # End of the current window
        current_window = Counter()  # Track words seen in the current window
        words_used = 0  # Count of valid words used
        
        while right + word_length <= len(s):  # Ensure we do not exceed the string length
            word = s[right:right + word_length]  # Extract the next word
            right += word_length  # Move the right pointer forward
            
            if word in word_freq:  # If it's a valid word
                current_window[word] += 1
                words_used += 1
                
                # If we have too many occurrences of a word, shrink the window from the left
                while current_window[word] > word_freq[word]:
                    leftmost_word = s[left:left + word_length]
                    left += word_length
                    current_window[leftmost_word] -= 1
                    words_used -= 1
                
                # If we have used all words exactly once, store the start index
                if words_used == word_count:
                    result.append(left)
            
            else:  # If an invalid word is found, reset window
                current_window.clear()
                words_used = 0
                left = right  # Move left pointer to restart from next valid position
    
    return result  # Return all valid starting indices
